### ISSUE TITLE
impersonation

### ISSUE THREAD

**Issue by dbinetti** [auth0-community/ember-simple-auth-auth0#107]
First, thanks SO MUCH for keeping this library going.  Those of us who are depending on you are grateful and appreciative.  

I've gotten V4.0.0.-beta.1 working in my main use case, which is wholly passwordless.  Huzzah!  

However, impersonation no longer works. Instead, I get a "state" error as listed below.

The Auth0 docs leave a lot to be desired on this end.  Searching through the library code doesn't seem to even have reference to the state parameter.  So I'm a bit stuck.  I'm leaving the "state" field blank in my impersonation request as I did under 3.0, but this doesn't appear to work in v4.  Thoughts?  Approach?

```
"Error: Auth0 returned error `invalid_token`: `state` does not match.↵    at Auth0Error.EmberError (http://localhost:4200/assets/vendor.js:28661:25)↵    at new Auth0Error (http://localhost:4200/assets/vendor.js:171830:16)↵    at auth0.parseHash (http://localhost:4200/assets/vendor.js:171568:27)↵    at WebAuth.validateAuthenticationResponse (http://localhost:4200/assets/vendor.js:86120:13)↵    at WebAuth.parseHash (http://localhost:4200/assets/vendor.js:86093:16)↵    at RSVP.Promise (http://localhost:4200/assets/vendor.js:171566:15)↵    at initializePromise (http://localhost:4200/assets/vendor.js:65985:7)↵    at new Promise (http://localhost:4200/assets/vendor.js:66464:35)↵    at Class._getUrlHashData (http://localhost:4200/assets/vendor.js:171565:14)↵    at tryCatcher (http://localhost:4200/assets/vendor.js:65782:21)"
__proto__
:
EmberError
```

**Comment by XaserAcheron** [comment]
Glad to help! Bit of an odd set of circumstances that led me here, but here we are. :P

Poking at this a bit, Auth0's guide on [User Impersonation](https://auth0.com/docs/user-profile/user-impersonation) does warn against leaving the state field blank:
> State: The state is a required parameter and leaving it blank may lead to errors like `Impersonation - Bad mac.` 

...Having said that, though, I can reproduce the error regardless of whether or not I put something in the state field, so something's definitely goofy. I'll take a look.

**Comment by XaserAcheron** [comment]
Tracked this down; it's an [auth0.js change](https://github.com/auth0/auth0.js/issues/683). The short version is that impersonation now needs to be enabled via flag when calling a certain method (parseHash, which this addon uses).

Auth0 has stated this was motivated by security concerns (though they haven't elaborated just yet), so I'm going to follow suit and add an `enableImpersonation` flag to Ember's environment configuration so it's strictly an opt-in thing. That sound good?

It's a breaking change, but it's convenient that we're preparing for a major release anyway. ;)

**Comment by dbinetti** [comment]
Perfect

**Comment by XaserAcheron** [comment]
Just published `v4.0.0-beta.2` with the new flag. Set it in your app's `config/environment.js` and you should be good to go!

```js
// config/environment.js
module.exports = function(environment) {
  let ENV = {
    'ember-simple-auth': {
      authenticationRoute: 'login',
      auth0: {
        clientID: '1234',
        domain: 'my-company.auth0.com',
        logoutReturnToURL: '/logout',
        enableImpersonation: true
      }
    }
  };

  return ENV;
};
```
